text/0000-const-fn-in-trait.md

```diff
@@ -0,0 +1,148 @@
+- Feature Name: `const_fn_in_trait`
+- Start Date: 2023-09-15
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue:
+  [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+[summary]: #summary
+
+This feature allows marking functions in traits as `const`. Users of the trait
+will be able to use these functions in const contexts.
+
+# Motivation
+
+[motivation]: #motivation
+
+Currently, there is no implementation for high-level generic programming in
+const contexts. Typically traits provide generic interfaces in Rust, but they do
+not support `const` methods.
+
+This RFC proposes allowing functions in a trait to be marked `const`, meaning
+they can be called from a const context like other const functions. Use cases
+include:
+
+- Needing a constructor that can create a static object for C-style plugins
+- Wanting to reduce code duplication in const contexts
+- Subtraits that want to provide defaults based on supertraits
+
+Workarounds typically involve a combination of wrapper functions, macros, and
+associated consts. This RFC will eliminate the need for such workarounds.
```

Comment:
I feel it may be useful to note some of these workarounds?

Essentially, I'm searching for an illustration of what is possible (although maybe hugely inconvenient and ugly) in current Rust, and what new possibilities would be enabled by this proposal, if any?

Reply:
Do you mean that you are looking for a sample of what a solution might look like today? Or just a more in depth explanation from the list?

Reply:
Maybe just a better explanation of each workaround or the use cases would be best, current examples are usually pretty large and vary a lot

Reply:
Hmm, mostly I fear that there may be a lot of edge-cases that have not yet been considered in the proposal?

And then I think that if I saw a few instances of "this thing can already be done today, just not as cleanly", then it'd assuage my worries on that front.